compiler/luci/pass/src/RemoveUnnecessaryTransposeNetPass.cpp

```diff
@@ -0,0 +1,453 @@
+  uint32_t old_shape_start_idx = 0;
+  uint32_t old_shape_end_idx = 1;
+  uint32_t old_shape_product = _shape[old_shape_start_idx].value;
+
+  auto expand_range = [&]() -> bool {
+    if (old_shape_end_idx >= _shape.size())
+      return false;
+
+    old_shape_product *= _shape[old_shape_end_idx].value;
+    old_shape_end_idx++;
+    return true;
+  };
+
+  auto move_to_next_range = [&]() -> bool {
+    if (old_shape_end_idx >= _shape.size())
+      return false;
+
+    old_shape_start_idx = old_shape_end_idx;
+    old_shape_end_idx++;
+    old_shape_product = _shape[old_shape_start_idx].value;
+    return true;
+  };
+
+  // Add tags from '_shape' to the 'new_shape'
+  uint32_t new_shape_idx = 0;
+  while (new_shape_idx < new_shape.size())
+  {
+    Dim &target_dim = new_shape[new_shape_idx];
+
+    // Ignore dim == 1
+    if (target_dim.value == 1)
+    {
+      new_shape_idx++;
+      continue;
+    }
+
+    while (old_shape_product < target_dim.value)
+    {
+      if (expand_range() == false)
+        break;
+    }
+
+    if (old_shape_product != target_dim.value)
+      return false;
+
+    // old_shape_product == target_dim.value
+    for (uint32_t idx = old_shape_start_idx; idx < old_shape_end_idx; idx++)
+    {
+      const auto &old_tags = _shape[idx].tags;
+      target_dim.tags.insert(target_dim.tags.end(), old_tags.begin(), old_tags.end());
+    }
+
+    new_shape_idx++;
+    move_to_next_range();
+  }
+  _shape = new_shape;
+  return true;
+}
+
+bool TaggedShapeAnalyzer::verify_tag() const
+{
+  // check whether tags in _shape are incremental
+  uint8_t tag = START_TAG;
+  for (const auto &dim : _shape)
+  {
+    for (const auto &t : dim.tags)
+    {
+      if (t == tag)
+        tag++;
+      else
+        return false;
+    }
+  }
+  return true;
+}
+
+template <loco::DataType DType>
+bool TaggedShapeAnalyzer::can_remove_transposes(const luci::CircleTranspose *f_tr,
+                                                const luci::CircleReshape *m_rs,
+                                                const luci::CircleTranspose *b_tr)
+{
+  assert(loco::must_cast<luci::CircleConst *>(f_tr->perm())->dtype() == DType);
+  assert(loco::must_cast<luci::CircleConst *>(m_rs->shape())->dtype() == DType);
+  assert(loco::must_cast<luci::CircleConst *>(b_tr->perm())->dtype() == DType);
+
+  const luci::CircleNode *in_tensor = loco::must_cast<luci::CircleNode *>(f_tr->a());
+
+  init_shape_with_tag(in_tensor);
+
+  if (not analyze_transpose<DType>(f_tr))
+    return false;
+
+  if (not analyze_reshape<DType>(m_rs))
+    return false;
+
+  if (not analyze_transpose<DType>(b_tr))
+    return false;
+
+  if (not verify_tag())
+    return false;
+
+  return true;
+}
+
+/**
+ * @brief create CircleReshape node that reshapes 'front_transpose input tensor shape' into
+ * 'back_transposes output tensor shape'
+ */
+template <loco::DataType ShapeType>
+luci::CircleReshape *create_reshape_node(loco::Graph *graph,
```

Comment:
If `graph` argument is needed to get its graph, the graph can be gotten from `graph` api with other circle nodes like below.
```cxx
auto g = front_transpose->graph();
```

Reply:
Ah, I explicit define `graph` as argument. 
Because as i understood, It isn't possible to get a `graph` from the `const` circle node. 

Since this function doesn't change any given CircleNode, I hope other circle nodes keep as const arguments.